File: samples/client/3_1_0_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/AdditionalPropertiesValidator.java
package org.openapijsonschematools.client.schemas.validation;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class AdditionalPropertiesValidator implements KeywordValidator {
    public final Class<? extends JsonSchema> additionalProperties;

    public AdditionalPropertiesValidator(Class<? extends JsonSchema> additionalProperties) {
        this.additionalProperties = additionalProperties;
    }

    @Override
    public @Nullable PathToSchemasMap validate(
        JsonSchema schema,
        @Nullable Object arg,
        ValidationMetadata validationMetadata,
        @Nullable List<PathToSchemasMap> containsPathToSchemas,
        @Nullable PathToSchemasMap patternPropertiesPathToSchemas,
        @Nullable PathToSchemasMap ifPathToSchemas
    ) {
        if (!(arg instanceof Map<?, ?> mapArg)) {
            return null;
        }
        Set<String> presentAdditionalProperties = new LinkedHashSet<>();
        for (Object key: mapArg.keySet()) {
            if (key instanceof String) {
                presentAdditionalProperties.add((String) key);
            }
        }
        if (schema.properties != null) {
            presentAdditionalProperties.removeAll(schema.properties.keySet());
        }
        PathToSchemasMap pathToSchemas = new PathToSchemasMap();
        for(String addPropName: presentAdditionalProperties) {
            @Nullable Object propValue = mapArg.get(addPropName);
            List<Object> propPathToItem = new ArrayList<>(validationMetadata.pathToItem());
            propPathToItem.add(addPropName);
            if (patternPropertiesPathToSchemas != null && patternPropertiesPathToSchemas.containsKey(propPathToItem)) {
                continue;
            }
            ValidationMetadata propValidationMetadata = new ValidationMetadata(
                    propPathToItem,
                    validationMetadata.configuration(),
                    validationMetadata.validatedPathToSchemas(),
                    validationMetadata.seenClasses()
            );
            JsonSchema addPropsSchema = JsonSchemaFactory.getInstance(additionalProperties);
            if (propValidationMetadata.validationRanEarlier(addPropsSchema)) {
                // todo add_deeper_validated_schemas
                continue;
            }
            PathToSchemasMap otherPathToSchemas = JsonSchema.validate(addPropsSchema, propValue, propValidationMetadata);
            pathToSchemas.update(otherPathToSchemas);
        }
        return pathToSchemas;
    }
}

File: samples/client/3_1_0_unit_test/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/AdditionalPropertiesValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class AdditionalPropertiesValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        if (data.arg !is Map<*, *>) {
            return null
        }
        val additionalProperties: Class<out JsonSchema<*>> = data.schema.additionalProperties ?: return null
        val presentAdditionalProperties: MutableSet<String> = LinkedHashSet()
        for (key in data.arg.keys) {
            if (key is String) {
                presentAdditionalProperties.add(key)
            }
        }
        val properties: Map<String, Class<out JsonSchema<*>>>? = data.schema.properties
        if (properties != null) {
            presentAdditionalProperties.removeAll(properties.keys)
        }
        val pathToSchemas = PathToSchemasMap()
        for (addPropName in presentAdditionalProperties) {
            val propValue: Any? = data.arg.get(addPropName)
            val propPathToItem: List<Any> = data.validationMetadata.pathToItem + addPropName
            if (data.patternPropertiesPathToSchemas != null && data.patternPropertiesPathToSchemas
                    .containsKey(propPathToItem)
            ) {
                continue
            }
            val propValidationMetadata = ValidationMetadata(
                propPathToItem,
                data.validationMetadata.configuration,
                data.validationMetadata.validatedPathToSchemas,
                data.validationMetadata.seenClasses
            )
            val addPropsSchema = JsonSchemaFactory.getInstance(additionalProperties)
            if (propValidationMetadata.validationRanEarlier(addPropsSchema)) {
                // todo add_deeper_validated_schemas
                continue
            }
            val otherPathToSchemas = JsonSchema.validate(addPropsSchema, propValue, propValidationMetadata)
            pathToSchemas.update(otherPathToSchemas)
        }
        return pathToSchemas
    }
}
